File: Harish Kumar/Dynamic Programming & Recursion/Number of Dice Rolls With Target Sum/Number of Dice Rolls With Target Sum.java
class Solution {
    
    private int gateWays(int d, int f, int target, int dp[][]){
    
    if(d == 0 && target == 0)
    {
        return 1;
    }
    else if(d < 0 || target < 0)
    {
        return 0;
    }
    else if(dp[d][target] != -1)
    {
        return dp[d][target];
    }
    
    int count = 0;
    
    for(int face = 1; face <= f;face++)
    {
        count += gateWays(d-1,f,target -face , dp);
        count = count % 1000000007;
    }
    
    return dp[d][target] = count;
}
    
    public int numRollsToTarget(int d, int f, int target) {
        
        int dp[][] = new int[d+1][target+1];
        
        for(int i = 0; i<= d;i++)
        {
            Arrays.fill(dp[i],-1);
        }
       
        return  gateWays(d,f,target,dp);
        
    }
}

File: Shreya/Recursion and DP/Number of Dice Rolls With Target Sum.cpp
class Solution {
public:
    int func(int d, int f, int target,vector<vector<int>>&dp)
    {
        if(d==0 && target!=0)
        {
            return 0;
        }
        if(d==0 && target==0)
        {
            return 1;
        }
        if(d!=0 && (target<0||target==0))
        {
            return 0;
        }
        if(dp[d][target]!=-1)
        {
            return dp[d][target];
        }
        int ans = 0;
        for(int i=1; i<=f; i++)
        {
            ans += func(d-1,f,target-i,dp); 
            ans = ans % 1000000007;
        }
        dp[d][target]=ans;
        return ans % 1000000007;
        
    }
    int numRollsToTarget(int d, int f, int target) 
    {
        vector<vector<int>>dp(d+1,vector<int>(target+1,-1));
        int num = func(d,f,target,dp);
        return num;
        
    }
};